### PR TITLE
Add manager role

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Adds manager role

Closes: #345 

## How this works

Adds a new role that can change enabled addresses, permission to only modify enabled addresses. 

* Manager Role cannot modify other managers + admins
* Manager Role is not activated until DUpgrade
* Manager role is also enabled.
* Manager roles in upgrades/genesis are not applied until DUpgrade
  * it only applies if AllowList configuration with timestamp T applied after DUpgrade. I.e if T < D, configured addresses won't have manager role *even after* Dupgrade, since it was configured before (T < D) DUpgrade and we don't reconfigure.
  * If allow list configuration is activated after DUpgrade (T >= D) and contains manager roles for addresses, those will be applied and addresses will become managers.
 * Existing precompiles will break because we have changed the `IsEnabled` signature to check if manager role is activated.
- [x] change precompile template 

## How this was tested

Added UTs

- [x] test locally with real network

## How is this documented

- [ ] Add documentation to docs repo

### CODE REVIEW

File: precompile/allowlist/role.go [hunk 1]
```diff
@@ -5,22 +5,29 @@ package allowlist
...
+var (
+	// NoRole - this is equivalent to common.Hash{} and deletes the key from the DB when set.
+	NoRole = Role(common.BigToHash(common.Big0))
+	// EnabledRole - allowed to call the precompile.
+	EnabledRole = Role(common.BigToHash(common.Big1))
+	// Admin - allowed to both modify the allowlist and call the precompile.
+	AdminRole = Role(common.BigToHash(common.Big2))
+	// Manager - allowed to add and remove only enabled address, without being able to call the precompile (i.e not enabled). Activated only after DUpgrade.
+	// We don't specifically give enabled role to the manager, because we want to keep the backward compatibility with IsEnabled() function.
+	// TODO: we can give enabled role to the manager. For that we need to remove the IsEnabled(), so we can break other precompiles and force them to adopt.
```
Comment (ceyonur) [inline precompile/allowlist/role.go:21]: 👁️

File: precompile/contract/interfaces.go [hunk 2]
```diff
@@ -19,14 +19,17 @@ type StatefulPrecompiledContract interface {
...
+// ChainConfig defines an interface that provides information to a stateful precompile
 // about the chain configuration. The precompile can access this information to initialize
 // its state.
 type ChainConfig interface {
 	// GetFeeConfig returns the original FeeConfig that was set in the genesis.
 	GetFeeConfig() commontype.FeeConfig
 	// AllowedFeeRecipients returns true if fee recipients are allowed in the genesis.
 	AllowedFeeRecipients() bool
+	// IsDUpgrade returns true if the time is after the DUpgrade.
+	// TODO: Should we get a more general function like "AvalancheRules"?
```
Comment (ceyonur) [inline precompile/contract/interfaces.go:31]: 👁️

File: precompile/allowlist/role.go (hunk 1 above)
Comment (ceyonur) [reply precompile/allowlist/role.go:21]: I guess we have to give enabled to the manager because `TxAllowList` checks if the role is enabled to accept a tx. Meaning that even if manager role can alter someone else's enabled role, they won't be able to issue the tx (because their tx will be rejected by `preCheck` in [state_transition](https://github.com/ava-labs/subnet-evm/blob/master/core/state_transition.go#L350))

File: plugin/evm/vm.go [hunk 3]
```diff
@@ -294,7 +295,10 @@ func (vm *VM) Initialize(
 
 	// We enforce network upgrades here, regardless of the chain config
 	// provided in the genesis file.
-	g.Config.MandatoryNetworkUpgrades = params.GetMandatoryNetworkUpgrades(chainCtx.NetworkID)
+	// Skip if we are running unit tests.
+	if chainCtx.NetworkID != avagoconstants.UnitTestID {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply plugin/evm/vm.go]: No need to special case this anymore, I gave a non-zero timestamp to DUpgrade in unit test networks. The timestamp is still in past (in 2020), but VM can alter it's time to deactivate/activate the DUpgrade.

File: precompile/allowlist/allowlist.go [hunk 4]
```diff
@@ -119,11 +123,20 @@ func createAllowListRoleSetter(precompileAddr common.Address, role Role) contrac
...
 		// Verify that the caller is an admin with permission to modify the allow list
 		callerStatus := GetAllowListStatus(stateDB, precompileAddr, callerAddr)
-		if !callerStatus.IsAdmin() {
+		// At this point we know that Manager role is activated.
+		if IsManagerRoleActivated(evm) && callerStatus == ManagerRole {
+			// Get current status.
+			// Before the manager role, we never checked the status of the address we are trying to modify.
+			// So we should keep the same behaviour by special casing this.
+			modifyStatus := GetAllowListStatus(stateDB, precompileAddr, modifyAddress)
+			if !callerStatus.CanModify(IsManagerRoleActivated(evm), modifyStatus, role) {
```
Comment (darioush) [inline precompile/allowlist/allowlist.go:132]: nit: I think if we are only calling `CanModify` here inside the check for the manager role (to avoid reading the stateDB for the previous role, I assume), should we call it `CanManagerModify` and remove the other clauses from that function? I think otherwise it may give the impression that `CanModify` controls the behavior for other roles as well.
Also it is not needed to call `IsManagerRoleActivated` again. Possibly consider inlining the function.

File: precompile/contracts/nativeminter/config_test.go [hunk 5]
```diff
@@ -17,29 +17,30 @@ import (
...
 			ExpectedError: "cannot set address",
 		},
 		"duplicate admins in config in native minter allowlist": {
-			Config:        NewConfig(utils.NewUint64(3), append(admins, admins[0]), enableds, nil),
+			Config:        NewConfig(utils.NewUint64(3), append(admins, admins[0]), enableds, managers, nil),
 			ExpectedError: "duplicate address",
 		},
 		"duplicate enableds in config in native minter allowlist": {
-			Config:        NewConfig(utils.NewUint64(3), admins, append(enableds, enableds[0]), nil),
+			Config:        NewConfig(utils.NewUint64(3), admins, append(enableds, enableds[0]), managers, nil),
```
Comment (darioush) [inline precompile/contracts/nativeminter/config_test.go:31]: did we add a UT to check duplicate addresses in the manager list with the other lists?

File: precompile/contracts/nativeminter/contract_test.go [hunk 6]
```diff
@@ -58,6 +60,37 @@ var tests = map[string]testutils.PrecompileTest{
 			require.Equal(t, common.Big2, state.GetBalance(allowlist.TestEnabledAddr), "expected minted funds")
 		},
 	},
+	"mint funds from manager role role fails before activation": {
```
Comment (darioush) [inline precompile/contracts/nativeminter/contract_test.go:63]:
```suggestion
	"mint funds from manager role fails before activation": {
```

File: precompile/contracts/rewardmanager/contract_test.go [hunk 7]
```diff
@@ -93,6 +95,104 @@ var (
 				require.False(t, isFeeRecipients)
 			},
 		},
+		"set allow fee recipients from enabled succeeds after activation": {
```
Comment (darioush) [inline precompile/contracts/rewardmanager/contract_test.go:98]:
```suggestion
		"set allow fee recipients from manager succeeds after activation": {
```

File: precompile/testutils/test_precompile.go [hunk 8]
```diff
@@ -85,8 +88,13 @@ func (test PrecompileTest) setup(t testing.TB, module modules.Module, state cont
 	}
 
 	blockContext := contract.NewMockBlockContext(big.NewInt(test.BlockNumber), 0)
-	accesibleState := contract.NewMockAccessibleState(state, blockContext, snow.DefaultContextTest())
-	chainConfig := contract.NewMockChainState(commontype.ValidTestFeeConfig, false)
+	chainConfig := test.ChainConfig
+	if chainConfig == nil {
+		// DUpgrade is activated by default
+		chainConfig = contract.NewMockChainConfig(commontype.ValidTestFeeConfig, false, utils.NewUint64(0))
```
Comment (darioush) [inline precompile/testutils/test_precompile.go:94]: kinda nitpicky, but it would be nice to explicitly specify this instead of modifying it here.

File: precompile/allowlist/test_allowlist_config.go [hunk 9]
```diff
@@ -62,23 +111,63 @@ func AllowListConfigVerifyTests(module modules.Module) map[string]testutils.Conf
...
+			Config: mkConfigWithAllowList(module, &AllowListConfig{
+				AdminAddresses:   []common.Address{TestAdminAddr},
+				ManagerAddresses: []common.Address{TestManagerAddr},
+				EnabledAddresses: []common.Address{TestEnabledAddr},
+			}),
+			Other: mkConfigWithAllowList(module, &AllowListConfig{
+				AdminAddresses:   []common.Address{TestAdminAddr},
+				ManagerAddresses: []common.Address{TestManagerAddr},
+				EnabledAddresses: []common.Address{{3}},
+			}), Expected: false,
```
Comment (darioush) [inline precompile/allowlist/test_allowlist_config.go:158]: nit:
```suggestion
			}), 
			Expected: false,
```

File: contracts/contracts/test/ExampleDeployerListTest.sol [hunk 10]
```diff
@@ -12,12 +12,14 @@ import "./AllowListTest.sol";
...
 contract ExampleDeployerListTest is AllowListTest {
+  address constant ENABLED_ADDRESS = 0x0Fa8EA536Be85F32724D57A37758761B86416123;
+
   IAllowList allowList = IAllowList(DEPLOYER_LIST);
   ExampleDeployerList private example;
 
   function setUp() public {
     example = new ExampleDeployerList();
-    allowList.setNone(OTHER_ADDRESS);
+    allowList.setNone(ENABLED_ADDRESS);
```
Comment (darioush) [inline contracts/contracts/test/ExampleDeployerListTest.sol:22]: is this needed because the other test does not do a proper tear down?
(not needed to fix here, but could we open an issue or followup PR?)

**Comment by darioush** [comment]
Couple high level thoughts in the context of this PR:
- How should we handle the upgrades/rules passing? Do we want to make it easier to modify precompile behavior in upgrades?
- It seems we have a lot of code duplication in tests. This is not something we need to fix in this PR, but it seems some tests should be more at the unit level so they can have less of a setup/teardown.

File: params/network_upgrades.go [hunk 11]
```diff
@@ -4,14 +4,18 @@
...
 	"github.com/ava-labs/subnet-evm/utils"
 )
 
 var (
+	defaultActivationTime = time.Date(2020, time.December, 5, 5, 0, 0, 0, time.UTC)
+
 	LocalNetworkUpgrades = MandatoryNetworkUpgrades{
 		SubnetEVMTimestamp: utils.NewUint64(0),
-		DUpgradeTimestamp:  utils.NewUint64(0),
+		DUpgradeTimestamp:  utils.TimeToNewUint64(defaultActivationTime),
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply params/network_upgrades.go:18]: using a 0 means we activate the upgrade at genesis (if genesis.timestamp == 0), a timestamp in the past means they will be activated at the first block.

[the remark this answers is not in the document]
Comment (darioush) [reply params/network_upgrades.go:18]: do we rely on this behavior in tests?

File: contracts/contracts/test/ExampleDeployerListTest.sol (hunk 10 above)
Comment (ceyonur) [reply contracts/contracts/test/ExampleDeployerListTest.sol:22]: I presume, yes.

**Comment by ceyonur** [comment]
> Couple high level thoughts in the context of this PR:
> 
>     * How should we handle the upgrades/rules passing? Do we want to make it easier to modify precompile behavior in upgrades?
> 

We have discussed with @aaronbuchwald and agreed on using the direct upgrade times in functions: https://github.com/ava-labs/subnet-evm/issues/692. 
Another alternative would be creating precompiles (like [here](https://github.com/ava-labs/subnet-evm/issues/692)) when we activate them and provide chainConfig/accessibleState to those functions, so `createPrecompile` functions can change created functions according to upgrade.(they can map signatures to different functions, add new functions, remove existing functions etc) like:
```
func createNativeMinterPrecompile(accessibleState contract.AccessibleState) contract.StatefulPrecompiledContract {
	enabledFuncs := allowlist.CreateAllowListFunctions(ContractAddress)

	mintFunc := contract.NewStatefulPrecompileFunction(mintSignature, mintNativeCoin)

	funcs := append(enabledFuncs, mintFunc)

if accessibleState.GetChainConfig().IsXUpgrade(accessibleState.GetBlockContext().Timestamp()){
     newFunc := contract.NewStatefulPrecompileFunction(newSignature, newFunction)
     funcs = append(funcs, newFunc)
}
	// Construct the contract with no fallback function.
	contract, err := contract.NewStatefulPrecompileContract(nil, funcs)
	// TODO: Change this to be returned as an error after refactoring this precompile
	// to use the new precompile template.
	if err != nil {
		panic(err)
	}
	return contract
}
```

File: precompile/allowlist/role.go [hunk 12]
```diff
@@ -24,10 +24,12 @@ var (
 type Role common.Hash
 
 // IsNoRole returns true if [s] indicates no specific role.
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/allowlist/role.go]: good catch, changed.

File: contracts/test/tx_allow_list.ts [hunk 13]
```diff
@@ -68,4 +68,14 @@ describe("ExampleTxAllowList", function () {
...
+
+  test("should let manager to revoke allowed", "step_managerCanRevoke")
+
+  test("should let manager to add allowed", "step_managerCanAllow")
+
+  test("should not let manager to add admin", "step_managerCannotGrantAdmin")
+
+  test("should not let manager to revoke admin", "step_managerCannotRevokeAdmin")
+
+  test("should not let manager to grant admin", "step_managerCannotGrantAdmin")
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply contracts/test/tx_allow_list.ts]: removed

File: precompile/allowlist/role.go [hunk 14]
```diff
@@ -5,56 +5,92 @@ package allowlist
...
-func (s Role) IsEnabled() bool {
-	switch s {
-	case AdminRole, EnabledRole:
+func (r Role) CanModify(isManagerActivated bool, from, target Role) bool {
+	switch r {
+	case AdminRole:
 		return true
+	case ManagerRole:
+		return isManagerActivated &&
+			(from == EnabledRole && target == NoRole) || (from == NoRole && target == EnabledRole)
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/allowlist/role.go]: 🦅 👁️ 
added tests

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go [hunk 15]
```diff
@@ -69,6 +72,41 @@ import (
 				ExpectedErr: {{if $fail}} ErrCannot{{$func.Normalized.Name}}.Error() {{- else}} "" {{- end}},
 			},
 			{{- end}}
+			{{- if not $func.Original.IsConstant}}
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply accounts/abi/bind/precompilebind/precompile_contract_test_template.go:75]: this is not a view or constant function (i.e it can change the state)

File: accounts/abi/bind/precompilebind/precompile_contract_test_template.go [hunk 16]
```diff
@@ -69,6 +72,41 @@ import (
...
+			{{- range $isBefore := mkList true false}}
+			"calling {{decapitalise $func.Normalized.Name}} from Manager should {{- if $isBefore}} fail before activation {{- else}} succeed after activation{{- end}}":  {
+				Caller:     allowlist.TestManagerAddr,
+				ChainConfig: contract.NewMockChainConfig(commontype.ValidTestFeeConfig, false, {{- if $isBefore}} nil {{- else}} utils.NewUint64(0) {{- end}}),
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply accounts/abi/bind/precompilebind/precompile_contract_test_template.go:79]: I think we can do it with formatters.

File: plugin/evm/vm_test.go [hunk 17]
```diff
@@ -2197,18 +2195,47 @@ func TestBuildAllowListActivationBlock(t *testing.T) {
 // Test that the tx allow list allows whitelisted transactions and blocks non-whitelisted addresses
 func TestTxAllowListSuccessfulTx(t *testing.T) {
 	// Setup chain params
+	managerKey, err := crypto.HexToECDSA("8a1f9a8f95be41cd7ccb6168179afb4504aefe388d1e14474d32c45c72ce7b7b")
+	require.NoError(t, err)
+	managerAddress := crypto.PubkeyToAddress(managerKey.PublicKey)
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply plugin/evm/vm_test.go]: yea we can use `testEthAddrs[1]` (we have 2 addresses in that list). 
initially I had 3 different addreses but no need that; so we can use test addr here.

File: plugin/evm/vm_test.go [hunk 18]
```diff
@@ -2268,6 +2312,73 @@ func TestTxAllowListSuccessfulTx(t *testing.T) {
...
+	vm.clock.Set(reenableAllowlistTimestamp)
+
+	// Re-Submit a successful transaction
+	tx0 = types.NewTransaction(uint64(1), testEthAddrs[0], big.NewInt(1), 21000, big.NewInt(testMinGasPrice), nil)
+	signedTx0, err = types.SignTx(tx0, types.NewEIP155Signer(vm.chainConfig.ChainID), testKeys[0])
+	require.NoError(t, err)
+
+	errs = vm.txPool.AddRemotesSync([]*types.Transaction{signedTx0})
+	if err := errs[0]; err != nil {
+		t.Fatalf("Failed to add tx at index: %s", err)
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply plugin/evm/vm_test.go]: changed them, we should do a complete refactor on those though.

File: plugin/evm/vm_test.go [hunk 19]
```diff
@@ -2268,6 +2312,73 @@ func TestTxAllowListSuccessfulTx(t *testing.T) {
...
+	}
+
+	blk = issueAndAccept(t, issuer, vm)
+
+	// Verify that the constructed block only has the whitelisted tx
+	block = blk.(*chain.BlockWrapper).Block.(*Block).ethBlock
+
+	// Verify that etherBase has received fees
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply plugin/evm/vm_test.go]: 👍

File: plugin/evm/vm_test.go [hunk 20]
```diff
@@ -2268,6 +2312,73 @@ func TestTxAllowListSuccessfulTx(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that address 0 is whitelisted and address 1 is not
+	role = txallowlist.GetTxAllowListStatus(blkState, testEthAddrs[0])
+	if role != allowlist.AdminRole {
+		t.Fatalf("Expected allow list status to be set to admin: %s, but found: %s", allowlist.AdminRole, role)
+	}
+	// This is set as manager but should not be effectie until after the upgrade
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply plugin/evm/vm_test.go]: 👍

File: precompile/allowlist/allowlist.go (hunk 4 above)
Comment (ceyonur) [reply precompile/allowlist/allowlist.go]: I think it would be dangerous to provide the `CanModify` function without taking the `isManagerActivated` as we export those. if someone uses it before the activation that could potentially go wrong.

File: precompile/contracts/nativeminter/config_test.go (hunk 5 above)
Comment (ceyonur) [reply precompile/contracts/nativeminter/config_test.go:31]: yes we have added them in `test_allowlist_config.go`

File: precompile/contracts/nativeminter/contract_test.go (hunk 6 above)
Comment (ceyonur) [reply precompile/contracts/nativeminter/contract_test.go]: removed

File: precompile/testutils/test_precompile.go (hunk 8 above)
Comment (ceyonur) [reply precompile/testutils/test_precompile.go]: it would cause panics in precompile tests. added some comments for the default config.

File: precompile/allowlist/allowlist.go (hunk 4 above)
Comment (darioush) [reply precompile/allowlist/allowlist.go]: nit: seems we could assign mgrRoleActivated to a variable to avoid calling it twice or simply pass true then?

File: contracts/test/tx_allow_list.ts [hunk 21]
```diff
@@ -68,4 +68,12 @@ describe("ExampleTxAllowList", function () {
...
   test("should let admin to revoke allowed", "step_adminCanRevoke")
+
+  test("should let manager to revoke allowed", "step_managerCanRevoke")
+
+  test("should let manager to add allowed", "step_managerCanAllow")
+
+  test("should not let manager to add admin", "step_managerCannotGrantAdmin")
+
+  test("should not let manager to revoke admin", "step_managerCannotRevokeAdmin")
 })
```
Comment (anusha-ctrl) [inline contracts/test/tx_allow_list.ts:79]: should we add " should let manager deploy"

Comment (anusha-ctrl) [reply contracts/test/tx_allow_list.ts:79]: also manager cannot add another manager

File: precompile/allowlist/test_allowlist.go [hunk 22]
```diff
@@ -152,6 +155,264 @@ func AllowListTests(module modules.Module) map[string]testutils.PrecompileTest {
...
+			ReadOnly:    false,
+			ExpectedErr: ErrManagerCannotModify.Error(),
+		},
+		"set admin role to enabled from manager after activation": {
+			Caller:      TestManagerAddr,
+			BeforeHook:  SetDefaultRoles(contractAddress),
+			ChainConfig: contract.NewMockChainConfig(commontype.ValidTestFeeConfig, false, utils.NewUint64(0) /* activated */),
+
+			InputFn: func(t testing.TB) []byte {
+				input, err := PackModifyAllowList(TestAdminAddr, EnabledRole)
```
Comment (anusha-ctrl) [inline precompile/allowlist/test_allowlist.go:377]: great coverage

File: precompile/testutils/test_precompile.go [hunk 23]
```diff
@@ -85,8 +91,13 @@ func (test PrecompileTest) setup(t testing.TB, module modules.Module, state cont
...
 	blockContext := contract.NewMockBlockContext(big.NewInt(test.BlockNumber), 0)
-	accesibleState := contract.NewMockAccessibleState(state, blockContext, snow.DefaultContextTest())
-	chainConfig := contract.NewMockChainState(commontype.ValidTestFeeConfig, false)
+	chainConfig := test.ChainConfig
+	if chainConfig == nil {
+		// DUpgrade is activated by default
+		chainConfig = DefaultChainConfig
+	}
+
+	accesibleState := contract.NewMockAccessibleState(state, blockContext, snow.DefaultContextTest(), chainConfig)
```
Comment (anusha-ctrl) [inline precompile/testutils/test_precompile.go:100]: accesibleState -> accessibleState

File: precompile/allowlist/role.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (anusha-ctrl) [reply precompile/allowlist/role.go]: nice eye @darioush

**Review by anusha-ctrl — APPROVED** [review]
lgtm

File: params/network_upgrades.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ceyonur) [reply params/network_upgrades.go:18]: yes, we can set the initial vm clock before this; to have a deactivated upgrade.

File: precompile/allowlist/allowlist.go (hunk 4 above)
Comment (ceyonur) [reply precompile/allowlist/allowlist.go]: added `isManagerRoleActivated` variable on top of the function and used it here.

File: contracts/test/tx_allow_list.ts (hunk 21 above)
Comment (ceyonur) [reply contracts/test/tx_allow_list.ts:85]: added

File: contracts/contracts/test/ExampleDeployerListTest.sol (hunk 10 above)
Comment (richardpringle) [reply contracts/contracts/test/ExampleDeployerListTest.sol:22]: Making sure the state is what you want it to be in `setUp` is better than relying on teardown. I agree that we should also do teardown, but that shouldn't change the `setUp` function here.

File: contracts/contracts/test/ExampleTxAllowListTest.sol [hunk 24]
```diff
@@ -162,4 +164,140 @@ contract ExampleTxAllowListTest is AllowListTest {
...
+    assertTrue(!example.isEnabled(exampleAddress));
+    assertTrue(!example.isManager(managerAddress));
+
+    allowList.setManager(managerAddress);
+
+    assertTrue(manager.isManager(managerAddress));
+
+    manager.setEnabled(exampleAddress);
+    assertTrue(example.isEnabled(exampleAddress));
+  }
```
Comment (richardpringle) [inline contracts/contracts/test/ExampleTxAllowListTest.sol:183]: 👍

File: contracts/contracts/test/ExampleTxAllowListTest.sol [hunk 25]
```diff
@@ -162,4 +164,140 @@ contract ExampleTxAllowListTest is AllowListTest {
...
+
+    allowList.setEnabled(exampleAddress);
+    allowList.setManager(managerAddress);
+
+    assertTrue(example.isEnabled(exampleAddress));
+    assertTrue(manager.isManager(managerAddress));
+
+    manager.revoke(exampleAddress);
+    assertTrue(!example.isEnabled(exampleAddress));
+  }
```
Comment (richardpringle) [inline contracts/contracts/test/ExampleTxAllowListTest.sol:202]: 👍

File: contracts/contracts/test/ExampleTxAllowListTest.sol
Comment (richardpringle) [inline contracts/contracts/test/ExampleTxAllowListTest.sol:1]: Tests look pretty good overall! 🥇

File: contracts/README.md
Comment (richardpringle) [inline contracts/README.md:1]: Edits like this should be landed in a separate PR.

```bash
git checkout main
git pull
git checkout -b readme-fixup
git checkout add-manager-role -- contracts/README.md
git commit
gh pr create
```

File: contracts/contracts/test/ExampleDeployerListTest.sol
Comment (richardpringle) [inline contracts/contracts/test/ExampleDeployerListTest.sol:1]: Wait... should this be a separate change as well?

File: contracts/test/tx_allow_list.ts [hunk 26]
```diff
@@ -61,11 +61,25 @@ describe("ExampleTxAllowList", function () {
 
   test("should allow admin to add allowed address as allowed through contract", "step_enableThroughContract")
 
-  test("should let allowed address deploy", "step_canDeploy")
+  test("should let allowed address to send tx", "step_canDeploy")
```
Comment (richardpringle) [inline contracts/test/tx_allow_list.ts:64]:
```suggestion
  test("should let allowed-address send a tx", "step_canDeploy")
```

File: contracts/test/tx_allow_list.ts [hunk 27]
```diff
@@ -61,11 +61,25 @@ describe("ExampleTxAllowList", function () {
...
 
   test("should not let allowed add another allowed", "step_onlyAdminCanEnable")
 
   test("should not let allowed to revoke admin", "step_onlyAdminCanRevoke")
 
   test("should let admin to revoke allowed", "step_adminCanRevoke")
+
+  test("should let manager to revoke allowed", "step_managerCanRevoke")
```
Comment (richardpringle) [inline contracts/test/tx_allow_list.ts:72]:
```suggestion
  test("should let manager revoke allowed-address", "step_managerCanRevoke")
```

File: contracts/test/tx_allow_list.ts [hunk 28]
```diff
@@ -61,11 +61,25 @@ describe("ExampleTxAllowList", function () {
...
+
+  test("should let manager to add allowed", "step_managerCanAllow")
```
Comment (richardpringle) [inline contracts/test/tx_allow_list.ts:74]:
```suggestion
  test("should let manager add address to allow-list", "step_managerCanAllow")
```

File: contracts/test/tx_allow_list.ts [hunk 29]
```diff
@@ -61,11 +61,25 @@ describe("ExampleTxAllowList", function () {
...
+
+  test("should not let manager to add admin", "step_managerCannotGrantAdmin")
+
+  test("should not let manager to revoke admin", "step_managerCannotRevokeAdmin")
+
+  test("should not let manager to revoke manager", "step_managerCannotRevokeManager")
+
+  test("should not let manager to add manager", "step_managerCannotGrantManager")
```
Comment (richardpringle) [inline contracts/test/tx_allow_list.ts:82]:
```suggestion
  test("should not let manager add admin", "step_managerCannotGrantAdmin")

  test("should not let manager revoke admin", "step_managerCannotRevokeAdmin")

  test("should not let manager revoke manager", "step_managerCannotRevokeManager")

  test("should not let manager add manager", "step_managerCannotGrantManager")
```

File: params/network_upgrades.go
Comment (richardpringle) [inline params/network_upgrades.go:1]: Seems like some of these changes could have been split into a separate PR. Not saying you should necessarily do this now as there have been a few reviews already, but it shouldn't be too much more work to split out some of the functionality after you've implemented it (but before people review your code). It makes reviews easier (this PR is massive) and makes the diff easier to inspect when looking back (after your changes have landed).

File: plugin/evm/vm_test.go [hunk 30]
```diff
@@ -2967,28 +3078,6 @@ func TestSkipChainConfigCheckCompatible(t *testing.T) {
 	require.NoError(t, reinitVM.Shutdown(context.Background()))
 }
 
-func TestMandatoryUpgrades(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (richardpringle) [reply plugin/evm/vm_test.go:2970]: Change should be part of a separate PR

**Review by richardpringle — COMMENTED** [review]
Wow... did you think your diff would be so big when you first took on this task? 

Nothing major in my comments. Just some suggestions for next time. Take a look at the precompile comments, they are nits though. There are also some grammar errors in the JS/TS tests, I would recommend applying those suggestions or something similar. 

Good job though! 🏆

**Comment by ceyonur** [comment]
Closing in favor of https://github.com/ava-labs/subnet-evm/pull/801/
